### ISSUE TITLE
Detect name conflicts within a reactor in 

### ISSUE THREAD

**Issue by cmnrd** [lf-lang/lingua-franca#1973]
The following code is legal LF
```
reactor Foo(bar:int=2) {
    input bar:int
    state bar: int
    method bar(): int {==}
    reaction bar(bar)
}
```
but it causes compile errors in the C++ target and likely also other targets. The problem is, that we don't enforce uniqueness of names in the reactor scope. Currently, there is only a Xtext generated mechanism that detects if triggers have the same name, but it does not handle other reactor elements.

**Comment by lhstrh** [comment]
I have a fix for this, but it is incompatible with `test/Rust/src/CtorParamMixed.lf`:

```
target Rust

reactor Print(value: i32 = 42, name: String = {= "xxx".into() =}, other: bool = false) {
  state value = value
  state name = name
  state other = other

  reaction(startup) {=
    assert_eq!(42, self.value);
    assert_eq!("x2hr", self.name.as_str());
    assert_eq!(true, self.other);
    println!("success");
  =}
}

main reactor CtorParamMixed {
  p = new Print(other=true, name = {= "x2hr".into() =})
}
```
To fix this test, the constructor parameters would have to be renamed... Do we want this?

**Comment by cmnrd** [comment]
Interesting. I am curious how the rust generator handles this. Is 'other' essentially handled like a mutable parameter?

I think the safe thing to do for now is enforce uniqueness of names, also between state variables and parameters in all targets. If we think we need "mutable parameters", then we can think about how to express them nicely in LF and properly introduce them in all targets. But I would like to hear @oowekyala's opinion also.

**Comment by lhstrh** [comment]
@oowekyala?

**Comment by lhstrh** [comment]
> IMHO unless there is a technical reason to forbid conflicts we shouldn't forbid them. This kind of naming constraint doesn't really improve your code, but can be frustrating to the programmer, especially when they feel arbitrary. Just my 2 cents anyway

I agree. I noticed this when I tried to "fix" the test that started failing and had to rename the parameters...

I think the only way to address this in the C target would be to organize things in the self struct differently, e.g.: `self->state->foo` and `self->parm->foo` to distinguish between the state variable and the parameter. @edwardalee: WDYT?

**Comment by edwardalee** [comment]
I think it's bad enough in the C target that we have to refer to a state variable (and parameter) as `self->name`.  To have to do `self->state->name` would be even worse.

Some time ago, I tried to change the C target so you could just use `name`.  I failed.  But that would make the name conflicts even worse.

**Comment by lhstrh** [comment]
It sounds like we should make the check target-specific then...

**Comment by cmnrd** [comment]
What @oowekyala is saying though, is that in the Rust target parameters do not become part of the self struct. They are only available during construction (passed as arguments to the constructor). If they are to be stored on the self struct, then they need to be explicitly assigned to a state variable (which can be constant or mutable). Since parameters are not stored on the self struct, they also don't need to be checked for bane conflicts.

This view makes a lot of sense to me, but bringing the other targets in line with this view on parameters would be a lot of work...

**Comment by edwardalee** [comment]
Does this mean that in Rust you cannot access parameter values in a reaction body?  That would be a rather severe limitation.

**Comment by oowekyala** [comment]
Yes that's exactly it. I don't think it is limiting as the programmer can just save the parameter into a field if they want. It's actually giving more freedom to the lf programmer.

**Comment by edwardalee** [comment]
I think we have quite a few issues that are more important than this. I think what we should do is that if _any_ target has a problem with name collisions, then we should forbid them in the validator for all targets and call it done. Later, if someone has the inclination and the time, they can reverse this restriction in a backward-compatible way.

**Comment by lhstrh** [comment]
So what is your take on https://github.com/lf-lang/lingua-franca/issues/1973#issuecomment-1793882099 then? Adopt that solution as-is and change the tests?

**Comment by lhstrh** [comment]
If it's not important, we can also just remove the issue from the milestone...

**Comment by edwardalee** [comment]
> So what is your take on [#1973 (comment)](https://github.com/lf-lang/lingua-franca/issues/1973#issuecomment-1793882099) then? Adopt that solution as-is and change the tests?

I think it's very odd that you can't access reactor parameters in reaction bodies in the Rust target. Why are we calling them reactor parameters then?  I would change the test and rename the parameter.

**Comment by lhstrh** [comment]
> > So what is your take on [#1973 (comment)](https://github.com/lf-lang/lingua-franca/issues/1973#issuecomment-1793882099) then? Adopt that solution as-is and change the tests?
> 
> I think it's very odd that you can't access reactor parameters in reaction bodies in the Rust target. Why are we calling them reactor parameters then? I would change the test and rename the parameter.

I don't think that's the case. @oowekyala wrote:
> Parameters in the Rust target are treated as constructor parameters, they are not implicitly stored as state variables, so they are not a mutable parameter here.

In Java, a constructor argument also isn't treated as a class variable. I think _that_ would be odd.

**Comment by edwardalee** [comment]
Hmm...  Is a reactor declaration a constructor declaration?

**Comment by lhstrh** [comment]
If it's not, what is it? In all targets with languages that have OOP features, parameters are mapped to constructor arguments. What other reasonable way is there?
